File: app/src/main/java/me/zhanghai/android/douya/ui/TimeTextView.java
/*
 * Copyright (c) 2016 Zhang Hai <Dreaming.in.Code.ZH@Gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.ui;

import android.annotation.TargetApi;
import android.content.Context;
import android.os.Build;
import android.util.AttributeSet;
import android.widget.TextView;

import org.threeten.bp.ZonedDateTime;
import org.threeten.bp.format.DateTimeParseException;

import me.zhanghai.android.douya.util.LogUtils;
import me.zhanghai.android.douya.util.TimeUtils;

public class TimeTextView extends TextView {

    private static final int UPDATE_TIME_TEXT_INTERVAL_MILLI = 30 * 1000;

    private final Runnable mUpdateTimeTextRunnable = new Runnable() {
        @Override
        public void run() {
            updateTimeText();
        }
    };

    private ZonedDateTime mTime;

    public TimeTextView(Context context) {
        super(context);
    }

    public TimeTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public TimeTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public TimeTextView(Context context, AttributeSet attrs, int defStyleAttr,
                        int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    public ZonedDateTime getTime() {
        return mTime;
    }

    public void setTime(ZonedDateTime time) {
        mTime = time;
        updateTimeText();
    }

    /**
     * Should behave the same as {@link TimeUtils#formatDoubanDateTime(String, Context)}.
     */
    public void setDoubanTime(String doubanTime) {
        try {
            setTime(TimeUtils.parseDoubanDateTime(doubanTime));
        } catch (DateTimeParseException e) {
            LogUtils.e("Unable to parse date time: " + doubanTime);
            e.printStackTrace();
            setText(doubanTime);
        }
    }

    private void updateTimeText() {
        removeCallbacks(mUpdateTimeTextRunnable);
        if (mTime != null) {
            setTimeText(formatTime(mTime));
            postDelayed(mUpdateTimeTextRunnable, UPDATE_TIME_TEXT_INTERVAL_MILLI);
        }
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();

        if (mTime != null) {
            updateTimeText();
        }
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();

        removeCallbacks(mUpdateTimeTextRunnable);
    }

    protected String formatTime(ZonedDateTime time) {
        return TimeUtils.formatDateTime(time, getContext());
    }

    protected void setTimeText(String timeText) {
        setText(timeText);
    }
}

File: app/src/main/java/me/zhanghai/android/douya/ui/TimeTextView.kt
/*
 * Copyright (c) 2020 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.douya.ui

import android.content.Context
import android.util.AttributeSet
import com.google.android.material.textview.MaterialTextView
import me.zhanghai.android.douya.util.formatHumanFriendly
import org.threeten.bp.Duration
import org.threeten.bp.ZonedDateTime

class TimeTextView : MaterialTextView {
    companion object {
        private val UPDATE_TEXT_INTERVAL_MILLIS = Duration.ofSeconds(30).toMillis()
    }

    private val updateTextRunnable = Runnable { updateText() }

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context, attrs, defStyleAttr
    )

    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes)

    var time: ZonedDateTime? = null
        set(value) {
            field = value
            updateText()
        }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()

        if (time != null) {
            updateText()
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()

        removeCallbacks(updateTextRunnable)
    }

    private fun updateText() {
        removeCallbacks(updateTextRunnable)
        text = time?.formatHumanFriendly(context)
        if (time != null) {
            postDelayed(updateTextRunnable, UPDATE_TEXT_INTERVAL_MILLIS)
        }
    }
}
